### ISSUE TITLE
Quality assurance: cpplint errors

### ISSUE THREAD

**Issue by mweisgut** [hyrise/hyrise#2155]
Executing [`cpplint`](https://github.com/cpplint/cpplint) on Hyrise's source files (version e4a6d59ce) results in a long list of errors:
https://gist.githubusercontent.com/mweisgut/8044a3cbc91312f97b8a28a070908aeb/raw/b73232580c7371f6f4869ffc572f2129d354217f/gistfile1.txt

A more detailed analysis of these results will probably reveal some potential for improving the code quality:
- What errors are justified and should be addressed?
- How can those errors be avoided in the future, i.e., how can Hyrise's automated static code analysis, executed by the CI, be improved?

**Comment by Bouncner** [comment]
Just by skipping over it, it seems like it's mainly two issues:
 * non-const ref parameters: I googled that a lot since I am a C++ newbie, but from my understand (which comes mainly from this post: https://softwareengineering.stackexchange.com/a/299041), we're fine here and Google has simply a different style there
 * include_what_you_use: both @mrks and me played around with that as it is quite clear, that our includes are not very clean, but at least I was not happy with the results (but I don't know if they have improved over the past 1-2 years)

**Comment by mrks** [comment]
When grouping the output by warning types, it doesn't look that bad anymore:

```
[:~] $ curl https://gist.githubusercontent.com/mweisgut/8044a3cbc91312f97b8a28a070908aeb/raw/e154e50d0975b33ecc163cfdeff5bb24b1f80e93/gistfile1.txt | ggrep -oP '\[[a-z/_]*\]' | sort | uniq -c
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  205k  100  205k    0     0   622k      0 --:--:-- --:--:-- --:--:--  622k
   2 [build/header_guard]
   3 [build/include_order]
 784 [build/include_what_you_use]
   1 [build/namespaces]
 294 [runtime/references]
```

* build/header_guard, build/include_order, and build/namespaces come from boost_curry_override.hpp, which is somewhat outside of our control.
* build/include_what_you_use might be correct. When I spent tens of hours reducing the compile time, I also looked at where that time was going. The lion's share goes to template instantiations. Even when I wrote a script to remove unnecessary includes and replaced some includes with forward declarations, it didn't notably improve the compile time. While we should keep includes in mind, I think we can spent our time more productively.
* I disagree with runtime/references. Google seems to be overloading the pointer/reference semantics, and I am not quite sure why they are doing it.

**Comment by Bouncner** [comment]
> * I disagree with runtime/references. Google seems to be overloading the pointer/reference semantics, and I am not quite sure why they are doing it.

I think the stack overflow post nicely summarizes the idea behind Google's style guide and why it might no longer be that relevant with modern C++ versions.

**Comment by mweisgut** [comment]
>- build/include_what_you_use might be correct. When I spent tens of hours reducing the compile time, I also looked at where that time was going. The lion's share goes to template instantiations. Even when I wrote a script to remove unnecessary includes and replaced some includes with forward declarations, it didn't notably improve the compile time. While we should keep includes in mind, I think we can spent our time more productively.

I agree that we can spend our time more productively and should not create a task to fix all `build/include_what_you_use`. However, fixing these errors could be a long term side task, which could be done partially in each PR: we could check only modified .hpp and .cpp files with `cpplint` for `build/include_what_you_use` errors in the CI. Even if it's not the PR's author's fault, one error can be fixed by adding the missing include, i.e., one line. Therefore it's only a tiny extra effort to complete a PR.  Besides fixing these current errors incrementally over time, we would prevent the accidental omission of includes for newly implemented code.

If others also consider it useful, I would take care of it.

**Comment by mrks** [comment]
Maybe something to discuss in the next fortnightly. How would we deal with superfluous includes, which I consider the bigger issue?

**Comment by mweisgut** [comment]
Good point. Since `cpplint` also reports false positive `build/include_what_you_use` errors in some situations, having it as a hard CI barrier for a PR wouldn't be a good solution.

**Comment by mweisgut** [comment]
Bifortnightly discussion result: we should investigate which error types of cpplint/cppcheck/other tools are justified and create hard CI constraints for those.
